File: JAVA/final-BT-java/src/services/TransactionService.java
package services;

import config.DatabaseConnection;
import models.*;
import repositories.*;
import session.Session;

import java.math.BigDecimal;
import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;

public class TransactionService {

    private final Connection connection;

    private final TransactionRepository transactionRepository;
    private final UtilisateurRepository utilisateurRepository;
    private final BudgetRepository budgetRepository;
    private final CategorieRepository categorieRepository;

    public TransactionService() {
        connection = DatabaseConnection.getInstance().getConnection();
        transactionRepository = new TransactionRepository(connection);
        utilisateurRepository = new UtilisateurRepository(connection);
        budgetRepository = new BudgetRepository(connection);
        categorieRepository = new CategorieRepository(connection);
    }

    public boolean ajouter(Transaction transaction) {
        Utilisateur utilisateur = Session.getInstance().getUtilisateurConnecte();
        transaction.setIdUtilisateur(utilisateur.getIdUtilisateur());
        Categorie categorie = categorieRepository.rechercherParId(transaction.getIdCategorie());
        if (categorie == null) {
            return false;
        }
        transaction.setType(categorie.getType());
        if (transaction.getMontant().compareTo(BigDecimal.ZERO) <= 0) {
            System.out.println("\nLe montant doit être supérieur à 0.");
            return false;
        }
        if (categorie.getType() == TypeCategorie.DEPENSE
                && utilisateur.getSolde().compareTo(transaction.getMontant()) < 0) {
            System.out.println("\nSolde insuffisant.");
            return false;
        }
        try {
            connection.setAutoCommit(false);
            // ==========================
            // INSERT TRANSACTION
            // ==========================
            if (transactionRepository.ajouter(transaction) == 0) {
                connection.rollback();
                return false;
            }
            // ==========================
            // CALCUL DU NOUVEAU SOLDE
            // ==========================
            BigDecimal nouveauSolde;
            if (categorie.getType() == TypeCategorie.DEPENSE) {
                nouveauSolde = utilisateur.getSolde().subtract(transaction.getMontant());
            } else {
                nouveauSolde = utilisateur.getSolde().add(transaction.getMontant());
            }
            // ==========================
            // UPDATE UTILISATEUR
            // ==========================
            if (utilisateurRepository.modifierSolde(utilisateur.getIdUtilisateur(), nouveauSolde) == 0) {
                connection.rollback();
                return false;
            }
            // Mise à jour de la session
            utilisateur.setSolde(nouveauSolde);
            // ==========================
            // UPDATE BUDGET
            // ==========================
            if (categorie.getType() == TypeCategorie.DEPENSE) {
                Budget budget = budgetRepository.rechercherParCategorie(utilisateur.getIdUtilisateur(),
                        transaction.getIdCategorie());
                if (budget != null) {
                    BigDecimal nouveauConsomme = budget.getMontantConsomme().add(transaction.getMontant());
                    BigDecimal nouveauRestant = budget.getMontantMaximum().subtract(nouveauConsomme);
                    if (budgetRepository.modifierMontants(budget.getIdBudget(), nouveauConsomme, nouveauRestant) == 0) {
                        connection.rollback();
                        return false;
                    }
                    budget.setMontantConsomme(nouveauConsomme);
                    budget.setMontantRestant(nouveauRestant);
                    // ==========================
                    // ALERTE BUDGET
                    // ==========================
                    BigDecimal seuil = budget.getMontantMaximum().divide(new BigDecimal("5"));
                    if (nouveauRestant.compareTo(BigDecimal.ZERO) < 0) {
                        System.out.println("\nBUDGET DÉPASSÉ");
                        System.out.println("Catégorie : " + categorie.getNom());
                        System.out.println("Dépassement : " + nouveauRestant.abs() + " FCFA");
                    } else if (nouveauRestant.compareTo(seuil) <= 0) {
                        System.out.println("\n ATTENTION");
                        System.out.println("Le budget \"" + categorie.getNom() + "\" est presque épuisé.");
                        System.out.println("Montant restant : " + nouveauRestant + " FCFA");
                    }
                }
            }
            connection.commit();
            return true;
        } catch (Exception e) {
            try {
                connection.rollback();
            } catch (SQLException ex) {
                ex.printStackTrace();
            }
            e.printStackTrace();
            return false;
        } finally {
            try {
                connection.setAutoCommit(true);
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }

    public List<Transaction> lister() {

        int idUtilisateur = Session.getInstance()
                .getUtilisateurConnecte()
                .getIdUtilisateur();

        return transactionRepository.listerParUtilisateur(idUtilisateur);

    }

    public boolean modifier(Transaction transaction) {

        return transactionRepository.modifier(transaction) > 0;

    }

    public boolean supprimer(int idTransaction) {

        Transaction transaction = transactionRepository.rechercherParId(idTransaction);

        if (transaction == null) {
            return false;
        }

        return transactionRepository.supprimer(idTransaction) > 0;

    }
}

File: PHP/src/Services/TransactionService.php
<?php

namespace App\Services;

use App\Entity\Transaction;
use App\Repository\TransactionRepository;
use App\Repository\BudgetRepository;
use App\Repository\categorieRepository;
use App\Repository\UtilisateurRepository;
use App\Config\Database;


class TransactionService
{
    private TransactionRepository $transactionRepository;
    private BudgetRepository $budgetRepository;
    private CategorieRepository $categorieRepository;
    private UtilisateurRepository $utilisateurRepository;
    public function __construct()
    {
        $this->transactionRepository = new TransactionRepository();
        $this->budgetRepository = new BudgetRepository();
        $this->categorieRepository = new CategorieRepository();
        $this->utilisateurRepository = new UtilisateurRepository();
    }

    public function lister(int $idUtilisateur): array
    {
        return $this->transactionRepository->lister($idUtilisateur);
    }

    public function rechercherParId(int $id): ?Transaction
    {
        return $this->transactionRepository->rechercherParId($id);
    }

    public function ajouter(Transaction $transaction): bool
    {
        $pdo = Database::getInstance()->getConnection();
        $budget = null;
        try {
            $pdo->beginTransaction();
            $categorie = $this->categorieRepository->rechercherParId($transaction->getIdCategorie());
            if ($categorie == null) {
                $pdo->rollBack();
                return false;
            }
            $transaction->setType($categorie->getType());
            $utilisateur = $this->utilisateurRepository->rechercherParId($transaction->getIdUtilisateur());
            if ($utilisateur == null) {
                $pdo->rollBack();
                return false;
            }
            if ($categorie->getType() == "DEPENSE") {
                if ($utilisateur->getSolde() < $transaction->getMontant()) {
                    $pdo->rollBack();
                    return false;
                }
                $nouveauSolde = $utilisateur->getSolde() - $transaction->getMontant();
            } else {
                $nouveauSolde = $utilisateur->getSolde() + $transaction->getMontant();
            }
            if (!$this->transactionRepository->ajouter($transaction)) {
                $pdo->rollBack();
                return false;
            }
            if (!$this->utilisateurRepository->modifierSolde(
                $utilisateur->getIdUtilisateur(),
                $nouveauSolde
            )) {
                $pdo->rollBack();
                return false;
            }
            if ($categorie->getType() == "DEPENSE") {
                $budget = $this->budgetRepository->rechercherParCategorie($utilisateur->getIdUtilisateur(), $categorie->getIdCategorie());
                if ($budget != null) {
                    $budget->setMontantConsomme($budget->getMontantConsomme() + $transaction->getMontant());
                    $budget->setMontantRestant($budget->getMontantMaximum() - $budget->getMontantConsomme());
                    if (!$this->budgetRepository->modifierMontants($budget)) {
                        $pdo->rollBack();
                        return false;
                    }
                }
            }
            if ($budget != null) {
                if ($budget->getMontantRestant() <= 0) {
                    $_SESSION["warning"] = "⚠ Budget dépassé.";
                } elseif ($budget->getMontantRestant() <= ($budget->getMontantMaximum() * 0.20)) {
                    $_SESSION["warning"] = "⚠ Il reste moins de 20% du budget.";
                }
            }
            $pdo->commit();
            return true;
        } catch (\Exception $e) {
            $pdo->rollBack();
            return false;
        }
    }
    public function modifier(Transaction $transaction): bool
    {
        $pdo = Database::getInstance()->getConnection();
        try {
            $pdo->beginTransaction();
            $ancienneTransaction = $this->transactionRepository->rechercherParId($transaction->getIdTransaction());
            if ($ancienneTransaction == null) {
                $pdo->rollBack();
                return false;
            }
            $utilisateur = $this->utilisateurRepository->rechercherParId($ancienneTransaction->getIdUtilisateur());
            if ($utilisateur == null) {
                $pdo->rollBack();
                return false;
            }
            $categorie = $this->categorieRepository->rechercherParId($transaction->getIdCategorie());
            if ($categorie == null) {
                $pdo->rollBack();
                return false;
            }
            $transaction->setType($categorie->getType());
            $solde = $utilisateur->getSolde();
            if ($ancienneTransaction->getType() == "REVENU") {
                $solde -= $ancienneTransaction->getMontant();
            } else {
                $solde += $ancienneTransaction->getMontant();
            }
            $ancienBudget = $this->budgetRepository->rechercherParCategorie($utilisateur->getIdUtilisateur(),$ancienneTransaction->getIdCategorie());
            if ($ancienBudget != null && $ancienneTransaction->getType() == "DEPENSE") {
                if (!$this->budgetRepository->retirerConsommation($ancienBudget->getIdBudget(),$ancienneTransaction->getMontant())) {
                    $pdo->rollBack();
                    return false;
                }
            }
            if ($transaction->getType() == "REVENU") {
                $solde += $transaction->getMontant();
            } else {
                $solde -= $transaction->getMontant();
            }
            $nouveauBudget = $this->budgetRepository->rechercherParCategorie($utilisateur->getIdUtilisateur(),$transaction->getIdCategorie());
            if ($nouveauBudget != null && $transaction->getType() == "DEPENSE") {
                if (!$this->budgetRepository->ajouterConsommation($nouveauBudget->getIdBudget(),$transaction->getMontant())) {
                    $pdo->rollBack();
                    return false;
                }
            }
            if (!$this->transactionRepository->modifier($transaction)) {
                $pdo->rollBack();
                return false;
            }
            if (!$this->utilisateurRepository->modifierSolde($utilisateur->getIdUtilisateur(),$solde)) {
                $pdo->rollBack();
                return false;
            }
            $pdo->commit();
            return true;
        } catch (\Exception $e) {
            $pdo->rollBack();
            return false;
        }
    }
    public function supprimer(int $id): bool
    {
        $pdo = Database::getInstance()->getConnection();
        try {
            $pdo->beginTransaction();
            $transaction = $this->transactionRepository->rechercherParId($id);
            if ($transaction == null) {
                $pdo->rollBack();
                return false;
            }
            $utilisateur = $this->utilisateurRepository->rechercherParId($transaction->getIdUtilisateur());
            if ($utilisateur == null) {
                $pdo->rollBack();
                return false;
            }
            if ($transaction->getType() == "DEPENSE") {
                $budget = $this->budgetRepository->rechercherParCategorie($utilisateur->getIdUtilisateur(),$transaction->getIdCategorie());
                if ($budget != null) {$budget->setMontantConsomme($budget->getMontantConsomme()- $transaction->getMontant());
                    $budget->setMontantRestant($budget->getMontantMaximum() - $budget->getMontantConsomme());
                    if (!$this->budgetRepository->modifierMontants($budget)) {
                        $pdo->rollBack();
                        return false;
                    }
                }
            }
            $solde = $utilisateur->getSolde();
            if ($transaction->getType() == "REVENU") {
                $solde -= $transaction->getMontant();
            } else {
                $solde += $transaction->getMontant();
            }
            if (!$this->transactionRepository->supprimer($id)) {
                $pdo->rollBack();
                return false;
            }
            if (!$this->utilisateurRepository->modifierSolde($utilisateur->getIdUtilisateur(),$solde)) {
                $pdo->rollBack();
                return false;
            }
            $pdo->commit();
            return true;
        } catch (\Exception $e) {
            $pdo->rollBack();
            return false;
        }
    }
}
